### PR TITLE
Migrate TUMOnlineRequest to Retrofit

### PR DESCRIPTION
**Pull request by thellmund**
**This is work in progress**

This pull request replaces `TUMOnlineRequest` with `TUMOnlineClient`, which uses Retrofit. The steps we took: 
1. We replaced SimpleXML with TikXML due to improved performance, smaller memory footprint and cleaner API. 
1. We added `TUMOnlineClient` and `TUMOnlineAPIService` for API access. 
2. We added `TUMOnlineInterceptor` to intercept requests, append the token parameter and throw exceptions in case of errors. This is necessary because TUMonline always returns 200, even in case of errors.
3. We removed `TumLock` and its associated DAO.

Building on these changes, we should do these things next: 
1. Update string resources with more appropriate naming. 
2. Update design of error layouts (nicer icons, larger text, proper headers).
3. Add TypeConverters (e.g. convert date strings to DateTime objects).

Issues: #544 #674 #388

### CODE REVIEW

File: .idea/codeStyleSettings.xml [hunk 1]
```diff
@@ -1,264 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kordianbruck) [inline .idea/codeStyleSettings.xml]: Why were these files removed? They are in the repository on purpose.

Comment (thellmund) [reply .idea/codeStyleSettings.xml]: I did not actively remove them. No idea how that happened. I’ll make sure they are there again when I make the pull request

Comment (kordianbruck) [reply .idea/codeStyleSettings.xml]: That's usually why you check what files you are commiting, before doing the commit :clinking_glasses:

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+class TUMOnlineClient {
+
+    companion object {
+
+        private const val BASE_URL = "https://campus.tum.de/tumonline/"
+
+        private var apiService: TUMOnlineAPIService? = null
+
+        @JvmStatic fun getInstance(context: Context): TUMOnlineAPIService {
+            if (apiService == null) {
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt]: I think this method needs to be synchronized to avoid race conditions for two concurrent calls to `getInstance()`

**Comment by kordianbruck** [comment]
@thellmund this PR is getting huge. Lets merge this.

**Comment by thellmund** [comment]
@kordianbruck I’m aware, but that’s the nature of the beast. If something is used everywhere, changing it is a massive undertaking. That being said, I’m finishing the caching stuff and will hopefully be able to submit the pull request tomorrow.

**Comment by kordianbruck** [comment]
@thellmund still WIP or Ready to review?

**Comment by thellmund** [comment]
@kordianbruck Basically done, but I want to do some more testing before submitting it for review. Will likely do that tomorrow.

**Comment by thellmund** [comment]
@kordianbruck Ready for review

File: .idea/codeStyleSettings.xml (hunk 1 above)
Comment (pfent) [reply .idea/codeStyleSettings.xml]: @thellmund please revert changes to .idea files

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+import de.tum.`in`.tumcampusapp.component.tumui.tutionfees.model.TuitionList
+import de.tum.`in`.tumcampusapp.utils.CacheManager
+import de.tum.`in`.tumcampusapp.utils.Const
+import de.tum.`in`.tumcampusapp.utils.DateTimeUtils
+import retrofit2.Call
+import retrofit2.Retrofit
+
+class TUMOnlineClient(private val apiService: TUMOnlineAPIService) {
+
+    fun getCalendar(force: Boolean = false): Call<Events> {
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt]: I really don't like boolean parameters.
From looking at a function call, one can't tell WTH this means.

```
getCalendar(true);
```
Without looking at the function definition, that parameter could be anything.
Please consider a proper Enum as parameter, which might even directly reference the underlying String.

```
getCalendar(CacheControl.doNotCache)
```
Is much nicer to read

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CachingHelper.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        return response.newBuilder()
+                .removeHeader("Cache-Control")
+                .addHeader("Cache-Control", cacheControl.toString())
+                .build()
+    }
+
+    companion object {
+
+        private val cachingDurations = mapOf(
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CachingHelper.kt]: When we're not actually using the Map capabilities of that map, because we always linearly search the map, please use a List of tuples instead of a Map.

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CheckErrorInterceptor.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    private val tikXml = TikXml.Builder()
+            .exceptionOnUnreadXml(false)
+            .build()
+
+    @Throws(RequestLimitReachedException::class,
+            TokenLimitReachedException::class,
+            InvalidTokenException::class,
+            MissingPermissionException::class,
+            UnknownErrorException::class)
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CheckErrorInterceptor.kt]: If only there was a common exception type to be declared :thinking:

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/model/Error.kt [hunk 6]
```diff
@@ -1,7 +1,30 @@
...
+
+    val exception: IOException
+        get() = ERROR_MESSAGE_TO_EXCEPTION.entries
+                .filter { message.contains(it.key) }
+                .map { it.value }
+                .firstOrNull() ?: UnknownErrorException()
+
+    companion object {
+
+        private val ERROR_MESSAGE_TO_EXCEPTION = mapOf(
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/model/Error.kt]: Same as above.
Linear searching in a map is rather slow, so use a List instead

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarActivity.java [hunk 7]
```diff
@@ -406,72 +413,16 @@ public String interpretTime(int hour, int minutes) {
...
+    public void onEventDeleted(@NotNull String eventId) {
+        TcaDb.getInstance(this).calendarDao().delete(eventId);
+        refreshWeekView();
+        Utils.showToast(this, R.string.delete_event_confirmation);
     }
 
-    protected void editEvent(final CalendarItem calendarItem) {
+    @Override
+    public void onEditEvent(@NotNull CalendarItem calendarItem) {
+        // TODO: CalendarItem should implement Parcelable
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarActivity.java]: You still want to do this?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarController.java [hunk 8]
```diff
@@ -190,19 +190,20 @@ public CalendarItem getCalendarItemByStartAndEndTime(DateTime start, DateTime en
...
-            try {
-                replaceIntoDb(row);
-            } catch (Exception e) {
-                Utils.log(e);
+        List<Event> events = myCalendarList.getEvents();
+        if (events != null) {
+            for (Event event : events) {
+                // insert into database
+                try {
+                    replaceIntoDb(event);
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarController.java]: Ahhhh. Database transactions in a loop :unamused: 
Well, I think that's out of scope four your PR, oh well

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt [hunk 9]
```diff
@@ -66,20 +76,75 @@ class CalendarDetailsFragment : BottomSheetDialogFragment() {
...
+        AlertDialog.Builder(context)
+                .setTitle(R.string.event_delete_title)
+                .setMessage(R.string.delete_event_info)
+                .setPositiveButton(R.string.delete) { _, _ -> deleteEvent(eventId) }
+                .setNegativeButton(R.string.cancel, null)
+                .show()
+    }
+
+    private fun deleteEvent(eventId: String) {
+        context?.let { c ->
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt]: In those cases, I like a guarding elvis operator better.
```
val c = context ?: return
```

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt [hunk 10]
```diff
@@ -66,20 +76,75 @@ class CalendarDetailsFragment : BottomSheetDialogFragment() {
...
 
+    companion object {
+
+        @JvmStatic
+        fun newInstance(calendarItem: CalendarItem,
+                        listener: OnEventInteractionListener): CalendarDetailsFragment {
+            return CalendarDetailsFragment().apply {
+                this.arguments = Bundle().apply {
+                    putString(Const.CALENDAR_ID_PARAM, calendarItem.nr)
+                }
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt]: I think that second apply is unnecessary

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java [hunk 11]
```diff
@@ -182,67 +188,80 @@ private void updateDateViews() {
...
-            public void onFetch(DeleteEvent response) {
-                Utils.log("Event successfully deleted (now creating the edited version)");
-                TcaDb.getInstance(getApplicationContext())
-                     .calendarDao()
-                     .delete(eventNr);
-                createEvent();
-            }
+        // Because we don't show a loading screen for the delete request (only for the create
+        // request), we use a short Toast to let the user know that something is happening.
+        Toast.makeText(this, R.string.updating_event, Toast.LENGTH_SHORT).show();
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java]: Meh.
We should probably tackled this in another PR and show a snackbar

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java [hunk 12]
```diff
@@ -277,6 +296,8 @@ private void hideKeyboard() {
     private boolean handleOnBackPressed() {
         String title = titleView.getText().toString();
         String description = descriptionView.getText().toString();
+
+        // TODO: If the user is in edit mode, check whether any data was changed.
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java]: Do you still want to do this?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt [hunk 13]
```diff
@@ -15,34 +16,37 @@ import org.simpleframework.xml.Root
...
-                course.compareTo(other.course) * (-1)
-            } else {
-                headerId.compareTo(other.headerId) * (-1)
-            }
+    override fun compareTo(other: Exam): Int {
+        return if (this.semester == other.semester) {
+            course.compareTo(other.course) * (-1)
+        } else {
+            semester.compareTo(other.semester)
+        }
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt]: This should be [`compareBy()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/compare-by.html)

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java [hunk 14]
```diff
@@ -84,45 +64,66 @@ public void onCreate(Bundle savedInstanceState) {
...
+            // LectureAppointments need the name and id of the facing lecture
+            bundle.putString("stp_sp_nr", currentItem.getStp_sp_nr());
+            bundle.putString(Const.TITLE_EXTRA, currentItem.getTitle());
+
+            Intent i = new Intent(this, LecturesAppointmentsActivity.class);
+            i.putExtras(bundle);
+            startActivity(i);
+        });
+
+        mLectureId = getIntent().getStringExtra("stp_sp_nr");
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java]: that string needs to be extracted to a common constant

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java [hunk 15]
```diff
@@ -84,45 +64,66 @@ public void onCreate(Bundle savedInstanceState) {
...
-        tvLDetailsDozent.setText(currentItem.getVortragende_mitwirkende());
-        tvLDetailsOrg.setText(currentItem.getOrg_name_betreut());
-        tvLDetailsInhalt.setText(currentItem.getLehrinhalt());
-        tvLDetailsMethode.setText(currentItem.getLehrmethode());
-        tvLDetailsZiele.setText(currentItem.getLehrziel());
-        tvLDetailsLiteratur.setText(currentItem.getStudienbehelfe());
-        tvLDetailsTermin.setText(currentItem.getErsttermin());
-
-        showLoadingEnded();
+        tvLDetailsSWS.setText(String.format("%s - %s SWS",
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java]: that string lacks internationalization

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/adapter/LectureAppointmentsListAdapter.java [hunk 16]
```diff
@@ -74,50 +70,44 @@ public View getView(int position, View view, ViewGroup parent) {
...
+            // output if same day: we only show the date once
+            StringBuilder output = new StringBuilder();
+            if (DateTimeUtils.INSTANCE.isSameDay(start, end)) {
+                output.append(startDateOutput.print(start))
+                        .append(" - ")
+                        .append(endHoursOutput.print(end));
+            } else {
+                // show it normally
+                output.append(startDateOutput.print(start))
+                        .append("–")
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/adapter/LectureAppointmentsListAdapter.java]: This is an en-dash, while we used minus signs otherwise...
I think we should use en-dashes everywhere?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/adapter/LecturesListAdapter.java [hunk 17]
```diff
@@ -57,13 +53,14 @@ public View getView(int position, View view, ViewGroup parent) {
...
-        LecturesSearchRow lvItem = infoList.get(position);
+        Lecture lvItem = infoList.get(position);
 
         // if we have something to display - set for each lecture element
         if (lvItem != null) {
-            holder.tvLectureName.setText(lvItem.getTitel());
-            holder.tvTypeSWSSemester.setText(String.format("%s - %s - %s SWS", lvItem.getStp_lv_art_name(), lvItem.getSemester_id(), lvItem.getDauer_info()));
-            holder.tvDozent.setText(lvItem.getVortragende_mitwirkende());
+            holder.tvLectureName.setText(lvItem.getTitle());
+            holder.tvTypeSWSSemester.setText(String.format("%s - %s - %s SWS",
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/adapter/LecturesListAdapter.java]: i18n

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/Lecture.kt [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        @PropertyElement(name = "org_kennung_betreut") val chairTumId: String? = null,
+        @PropertyElement(name = "org_name_betreut") val chairName: String? = null,
+        @PropertyElement(name = "org_nr_betreut") val chairId: Int? = null,
+        @PropertyElement(name = "semester") val semester: String,
+        @PropertyElement(name = "semester_id") val semesterId: String,
+        @PropertyElement(name = "semester_name") val semesterName: String,
+        @PropertyElement(name = "sj_name") val semesterYears: String,
+        @PropertyElement(name = "stp_lv_art_kurz") val shortLectureType: String,
+        @PropertyElement(name = "stp_lv_art_name") val lectureType: String,
+        @PropertyElement(name = "stp_lv_nr") val stp_lv_nr: String,   // TODO: Rename valiables
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/Lecture.kt]: you still want to do this?

Comment (pfent) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/Lecture.kt]: At least fix the typo in the TODO :see_no_evil:

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        @PropertyElement(name = "org_kennung_betreut") val chairTumId: String? = null,
+        @PropertyElement(name = "org_name_betreut") val chairName: String? = null,
+        @PropertyElement(name = "org_nr_betreut") val chairId: String? = null,
+        @PropertyElement(name = "semester") val semester: String? = null,
+        @PropertyElement(name = "semester_id") val semesterId: String? = null,
+        @PropertyElement(name = "semester_name") val semesterName: String? = null,
+        @PropertyElement(name = "sj_name") val semesterYears: String? = null,
+        @PropertyElement(name = "stp_lv_art_kurz") val shortLectureType: String? = null,
+        @PropertyElement(name = "stp_lv_art_name") val lectureType: String? = null,
+        @PropertyElement(name = "stp_lv_nr") val stp_lv_nr: String,  // TODO: Rename variables
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt]: Leftover TODO

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+import com.tickaroo.tikxml.annotation.Xml
+
+/**
+ * This class is dealing with the deserialization of the output of TUMOnline to
+ * the method "DetailsLehrveranstaltungen".
+ */
+@Xml(name = "row")
+data class LectureDetails(
+        @PropertyElement(name = "dauer_info") val duration: String? = null,
+        @PropertyElement(name = "ersttermin") val firstAppointment: String? = null, // TODO: TypeConverters
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt]: Leftover TODO

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/person/model/Employee.kt [hunk 21]
```diff
@@ -66,21 +65,17 @@ data class Employee(@field:Element(name = "geschlecht", required = false)
...
     override fun toString(): String {
-        var infoText = title
-
-        if (title != "") {
-            infoText = title + ' '
-        }
-
-        return infoText + name + ' ' + surname
+        return arrayOf(title, name, surname)
+                .filterNotNull()
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/person/model/Employee.kt]: aren't you missing empty strings for titles then?

File: app/src/main/java/de/tum/in/tumcampusapp/api/app/Helper.java [hunk 22]
```diff
@@ -80,7 +80,7 @@ public static OkHttpClient getOkHttpClient(Context c) {
         builder.connectTimeout(Helper.HTTP_TIMEOUT, TimeUnit.MILLISECONDS);
         builder.readTimeout(Helper.HTTP_TIMEOUT, TimeUnit.MILLISECONDS);
 
-        builder.addNetworkInterceptor(new TumHttpLoggingInterceptor(message -> Utils.logwithTag(TAG, message)));
+        //builder.addNetworkInterceptor(new TumHttpLoggingInterceptor(message -> Utils.logwithTag(TAG, message)));
```
[the remark this answers is not in the document]
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/api/app/Helper.java]: Commented this for testing. Will re-enable.

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt (hunk 3 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/TUMOnlineClient.kt]: Android Studio displays the parameter name, but I agree with your feedback. Will adapt this accordingly.

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CachingHelper.kt (hunk 4 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CachingHelper.kt]: Done.

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CheckErrorInterceptor.kt (hunk 5 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/interceptors/CheckErrorInterceptor.kt]: 🤔

File: app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/model/Error.kt (hunk 6 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/api/tumonline/model/Error.kt]: Done.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarActivity.java (hunk 7 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarActivity.java]: That’s more of a “do this at some point” comment. People keep telling me my pull requests are too large 🙈

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarController.java (hunk 8 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarController.java]: More importantly, it’s not logic that I wrote. But now that you mention it, I can’t un-see it. Updated the DAO from `insert(CalendarItem)` to `insert(CalendarItem...)`.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt (hunk 9 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarDetailsFragment.kt]: You’re right, this is a little cleaner due to less indentation. Updated accordingly.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java (hunk 11 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java]: I agree. This is a quick fix for now, because we did not display any information before. 

I want to rethink how we show download statuses generally at some point. For instance, imagine we have a cached response that we display to the user. If the user decides to refresh and an exception occurs, we currently show a full-screen error. However, ideally, we would display it as a `Snackbar` or `AlertDialog` (depending on ceverity).

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java (hunk 12 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CreateEventActivity.java]: Again, a “do this at some point” comment.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt (hunk 13 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt]: Why?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java (hunk 14 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/activity/LecturesDetailsActivity.java]: Agreed. Can you (or @kordianbruck) tell me what this abbreviation stands for?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt (hunk 20 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/lectures/model/LectureDetails.kt]: Upon further inspection, I saw that the value can also be “geplant” (“planned”). For that reason, I’ll remove the comment.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/person/model/Employee.kt (hunk 21 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/person/model/Employee.kt]: Upon further inspection, I don’t believe that the string representation of `Employee` is used anywhere (we use `getNameWithTitle(Context)`). I’ve removed it now.

File: .idea/codeStyleSettings.xml (hunk 1 above)
Comment (thellmund) [reply .idea/codeStyleSettings.xml]: I know I forgot something... added them again.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt (hunk 13 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/model/Exam.kt]: I changed the sorting, but decided to use `compareByDescending()` and chain additional `thenBy()` calls.

File: app/src/main/res/values-de/strings.xml [hunk 23]
```diff
@@ -92,7 +83,7 @@
     <string name="licences_summary">Rechtliche Hinweise zu Lizenzen</string>
 
     <!-- Startup Wizard -->
-    <string name="token_not_enabled">Nicht angemeldet</string>
+    <string name="not_connected_to_tumonline">Nicht angemeldet</string>
```
Comment (pfent) [inline app/src/main/res/values-de/strings.xml]: This is not a proper translation of "Not connected"

File: app/src/main/java/de/tum/in/tumcampusapp/utils/StringExt.kt [hunk 24]
```diff
@@ -0,0 +1,9 @@
+package de.tum.`in`.tumcampusapp.utils
+
+fun String.compareTo(other: String?): Int {
+    return if (other != null) {
+        compareTo(other)
+    } else {
+        -1
+    }
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/StringExt.kt]: I think this is quite literally [`compareValues`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/compare-values.html)

File: app/src/main/java/de/tum/in/tumcampusapp/utils/CacheManager.kt [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+                    .execute()
+        } catch (e: IOException) {
+            Utils.log(e, "Error while loading calendar in CacheManager")
+        }
+    }
+
+    private fun syncPersonalLectures() {
+        TUMOnlineClient
+                .getInstance(context)
+                .getPersonalLectures(CacheControl.NO_CACHE)
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/CacheManager.kt]: It his correct?
We call this in `fillCache()`, but then set `NO_CACHE`?

Ok, I still think this is confusing...
This should probably be called `BYPASS_CACHE`

File: app/src/main/res/values-de/strings.xml (hunk 23 above)
Comment (thellmund) [reply app/src/main/res/values-de/strings.xml]: I think “angemeldet” fits better than “verbunden”, but since we seem to use the latter throughout the app, I’ve changed this translation.

File: app/src/main/java/de/tum/in/tumcampusapp/utils/CacheManager.kt (hunk 25 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/utils/CacheManager.kt]: I changed the naming accordingly and changed the method to `CacheControl.USE_CACHE`, because we only want to refresh the content in the background if it’s past its `max-age` (same behavior as before the refactoring).
